nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -645,8 +645,83 @@ impl StringPrototype {
         ))
     }
 
-    fn replace(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
+    /// ### [22.1.3.19 String.prototype.replace ( searchValue, replaceValue )](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.replace)
+    fn replace(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+
+        let search_value = args.get(0);
+        let replace_value = args.get(1);
+
+        // 2. If searchValue is neither undefined nor null, then
+        if !search_value.is_null() && !search_value.is_undefined() {
+            // a. Let replacer be ? GetMethod(searchValue, %Symbol.replace%).
+            let symbol = WellKnownSymbolIndexes::Replace.into();
+            let replacer = get_method(agent, search_value, symbol);
+
+            // b. If replacer is not undefined, Return ? Call(replacer, searchValue, « O, replaceValue »).
+            if let Ok(Some(replacer)) = replacer {
+                return call_function(
+                    agent,
+                    replacer,
+                    search_value,
+                    Some(ArgumentsList(&[o, replace_value])),
+                );
+            }
+        }
+
+        // 3. Let s be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        // 4. Let searchString be ? ToString(searchValue).
+        let search_string = to_string(agent, search_value)?;
+
+        // 5. Let functionalReplace be IsCallable(replaceValue).
+        if let Some(functional_replace) = is_callable(replace_value) {
+            // 7. Let searchLength be the length of searchString.
+            let search_length = search_string.len(agent);
+
+            // 8. Let position be StringIndexOf(s, searchString, 0).
+            let position = if let Some(position) = search_string.as_str(agent).find(s.as_str(agent))
+            {
+                position
+            } else {
+                // 9. If position is not-found, return s.
+                return Ok(s.into_value());
+            };
+
+            // 10. Let preceding be the substring of s from 0 to position.
+            // 11. Let following be the substring of s from position + searchLength.
+            // 12. If functionalReplace is true,
+            let preceding = &s.as_str(agent)[0..position].to_owned();
```

Comment:
**nitpick:** You can move these after the `call`; that avoids doing any allocations if the `functional_replace` throws an error.

Reply:
good point!